src/policy/ephemeral_policy.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include<policy/ephemeral_policy.h>
+#include<policy/policy.h>
+
+bool CheckValidEphemeralTx(const CTransactionRef tx, CFeeRate dust_relay_fee, CAmount base_fee, CAmount mod_fee, TxValidationState& state)
+{
+    // We never want to give incentives to mine this transaction alone
+    if ((base_fee != 0 || mod_fee != 0) &&
+        std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& output) { return IsDust(output, dust_relay_fee); })) {
+        return state.Invalid(TxValidationResult::TX_NOT_STANDARD, "dust", "tx with dust output must be 0-fee");
+    }
+
+    return true;
+}
+
+std::optional<Txid> CheckEphemeralSpends(const Package& package, CFeeRate dust_relay_rate, const CTxMemPool& tx_pool)
+{
+    if (!Assume(std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx != nullptr;}))) {
+        // Bail out of spend checks if caller gave us an invalid package
+        return std::nullopt;
+    }
+
+    std::map<Txid, CTransactionRef> map_txid_ref;
+    for (const auto& tx : package) {
+        map_txid_ref[tx->GetHash()] = tx;
+    }
+
+    for (const auto& tx : package) {
+        Txid txid = tx->GetHash();
+        std::unordered_set<COutPoint, SaltedOutpointHasher> unspent_parent_dust;
+
+        for (const auto& tx_input : tx->vin) {
+            // We look for an in-package or in-mempool dependency
+            CTransactionRef parent_ref = nullptr;
+            if (map_txid_ref.contains(tx_input.prevout.hash)) {
+                parent_ref = map_txid_ref[tx_input.prevout.hash];
+            } else {
+                parent_ref = tx_pool.get(tx_input.prevout.hash);
+            }
+
+            // Check for dust on parents
+            if (parent_ref) {
+                for (uint32_t out_index = 0; out_index < parent_ref->vout.size(); out_index++) {
+                    const auto& tx_output = parent_ref->vout[out_index];
+                    if (IsDust(tx_output, dust_relay_rate)) {
+                        unspent_parent_dust.insert(COutPoint(parent_ref->GetHash(), out_index));
+                    }
+                }
+            }
+        }
```

Comment:
Overall I think this is much easier to follow with the codepaths merged, thanks.  However on further review, I am concerned about how much work we might do in this function.

As written, if a transaction had 1000 inputs, which all spend from a single parent with 1000 outputs, then we'd do 1M loop iterations, because we're not deduplicating parents for each input of a transaction (so for each input of the child, we'd be looking at all 1000 outputs of the parent).  If we deduplicate the parents first, then the number of parent outputs we might possibly have to look at is bounded by the ancestor size limit (or the cluster size limit, in the future), to something on the order of a few thousand outputs. 

If we are relying on the ancestor size limits for a work bound, then we should make sure that all those limits are checked prior to this function being invoked.  I think in the package acceptance case, that would involve moving this check to come after the `CheckPackageLimits()` check that happens in `AcceptMultipleTransactions`.

I think this might be enough, but we could consider going further -- what if we were to cache how many dust outputs a transaction has in the `CTxMemPoolEntry` itself?  Then we would be reduced to just looping over the inputs of the incoming transactions and counting, for each distinct parent, how many of the dust outputs are spent. This is probably not simple to do at the moment (since in the package case, we don't have CTxMemPoolEntry objects to work on yet), but I have some refactors in mind for mempool acceptance which would make this easier to consider in the future.

Anyway, it might be nice to add a micro-benchmark for transaction validation of a single transaction with 1000 inputs, which all come from a single parent with 1000 outputs, just to make sure we're not slowing that case down inordinately.

Reply:
> we don't have CTxMemPoolEntry objects to work on yet

I don't think that's true? `PreChecks` has built a mempool entry for each package tx. We could track the exact outpoint that is dust in the mempool entry. I'll give that a shot.

Reply:
@sdaftuar this should do it if you think something like this is reasonable: https://github.com/instagibbs/bitcoin/commit/6007146613fe663f0af30c24c47a19792f8e691d

Reply:
Hm, I I think that approach uses an extra 24 bytes per mempool entry, which strikes me as a bit heavy for what we need.  Let's see if this is fast enough as-is and if so, call it a day.

Reply:
Alternative is to store a bool, and use that to filter 99.9% of transactions that won't have dust, and then eat the O(num_outputs) cost to find the dust.

With 1000 inputs/outputs and a small number of dust outputs, I get:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           60,600.53 |           16,501.51 |    0.7% |      0.01 | `MempoolCheckEphemeralSpends`
```

When I make all the outputs dust(which can't happen currently due to IsStandard):
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           75,160.69 |           13,304.83 |    0.6% |      1.07 | `MempoolCheckEphemeralSpends`

```

number of outputs/inputs can be varied in the benchmark via `-asymptote=X`

Reply:
I think this should be fine for performance.  Large transactions take orders of magnitude more time to validate than what is being added here (and for small transactions this is extremely fast).